src/helpers/hooks/useFormWithValidation.ts

```diff
@@ -1,77 +1,192 @@
+    value: T[K]
+    wasChanged: boolean
+    error?: string
+  }
+}
+
+type ContactFormFieldsState = FormFieldState<ContactFormFields>
+
+interface ContactFormState {
+  fields: ContactFormFieldsState
+  isSubmitting: boolean
+  formSuccessfullySent: boolean
+  submissionError?: string
+}
+
+type ContactFormAction<K extends keyof ContactFormFields = keyof ContactFormFields> =
+  | { type: 'UPDATE_FIELD'; payload: { field: K; value: ContactFormFields[K]; error?: string }}
+  | { type: 'START_FORM_SUBMISSION' }
+  | { type: 'SET_FORM_SUBMISSION_SUCCESS' }
+  | { type: 'SET_FORM_SUBMISSION_FAILURE'; payload: { error: string } }
+  | { type: 'COMPLETE_FORM_SUBMISSION' }
+
+const initialContactFormFieldsState: ContactFormFieldsState = {
+  name: { value: '', wasChanged: false },
+  email: { value: '', wasChanged: false },
+  message: { value: '', wasChanged: false },
+}
 
-const initialValues: IContactFormFields = {
-  name: '',
-  email: '',
-  message: '',
+const initialContactFormState: ContactFormState = {
+  fields: initialContactFormFieldsState,
+  isSubmitting: false,
+  formSuccessfullySent: false,
+}
+
+const reducer = (state: ContactFormState, action: ContactFormAction) => {
+  switch (action.type) {
+    case 'UPDATE_FIELD':
+      return {
+        ...state,
+        fields: {
+          ...state.fields,
+          [action.payload.field]: {
+            value: action.payload.value,
+            wasChanged: true,
+            error: action.payload.error || '',
+          },
+        },
+      }
+    case 'START_FORM_SUBMISSION':
+      return {
+        ...state,
+        isSubmitting: true,
+      }
+    case 'SET_FORM_SUBMISSION_SUCCESS':
+      return {
+        ...state,
+        fields: initialContactFormFieldsState,
+        isSubmitting: false,
+        formSuccessfullySent: true,
+      }
+    case 'SET_FORM_SUBMISSION_FAILURE':
+      return {
+        ...state,
+        isSubmitting: false,
+        submissionError: action.payload.error,
+      }
+    case 'COMPLETE_FORM_SUBMISSION':
+      return {
+        ...state,
+        formSuccessfullySent: false,
+      }
+    default:
+      return state
+  }
 }
 
 export const useFormWithValidation = () => {
-  const [values, setValues] = useState<IContactFormFields>(initialValues)
-  const [errors, setErrors] = useState<Partial<IContactFormFields>>({})
-  const [isValid, setIsValid] = useState(false)
-  const [isSubmitting, setIsSubmitting] = useState(false)
-  const [formSuccessfullySent, setFormSuccessfullySent] = useState(false)
-
-  const handleChange = ({
-    target,
-  }: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
-    const { name, value } = target
-
-    setValues({ ...values, [name]: value })
-    setErrors({ ...errors, [name]: target.validationMessage })
-    setIsValid(target.closest('form')!.checkValidity())
+  const [state, dispatch] = useReducer(reducer, initialContactFormState)
+
+  const getFieldError = <K extends keyof ContactFormFields>(
```

Comment:
@Yura33-dev As per the [Project Requirements](https://docs.google.com/document/d/1pxspVrS72lxQnVUIe7Fd8j-gBNpYT4vdSplA7ArCEhg/edit#heading=h.kctencwjew16), we also need to ensure that the `name` and `message` fields do not contain special characters like `<`, `>`, `&`, which can be interpreted as part of HTML code or JavaScript. This is important for preventing potential security vulnerabilities like injection attacks.

Please setup a another regex for that. We'll need to add `PATTERN_MISMATCH` messages for both the `name` and `message` fields in `CustomValidationMessages` to provide users with clear error message. For a name can use something like:
> Only letters, spaces, apostrophes ('), and hyphens (-) allowed

Please check with QA if you have any clarification on that 🙂

Reply:
Hi @Yura33-dev 🖐️ Thanks for adding pattern validation to the name and message fields! Would you consider these pattern mismatch error messages?

Name Field:
> Name can only include letters, spaces, and these: ' ’ -

Message Field:
> Yikes! Only letters, numbers, spaces, and ' ’ " , . ! ? - are allowed

Additionally, could you please check the message regex to make sure comma is listed only once?

Reply:
@Yura33-dev Did you see the message above?

Reply:
@naumch1k Yeep, I saw those messages a little later 🙈. But now I have added new commit with these enhancements 🙂

Reply:
@Yura33-dev I am sorry if I caused any confusion. I didn't mean to restrict multiple consecutive commas in `validMessageRegexp`. What I wanted was just to clean up a duplicate. 
<img width="680" alt="Screenshot 2024-09-30 at 11 49 34 AM" src="https://github.com/user-attachments/assets/ad8168a7-348d-4b44-a499-d2e10c730245">


Could you please explain the change you made to `validNameRegexp`? Particularly, the need in `z` after `-`?

Reply:
@naumch1k Ops! Okay, I have deleted one comma from `validMessageRegexp`. 
I suggest it looks like wrong (I mean `validNameRegexp` with `z`). I just deleted it from regexp. 🙂
I have pushed a new commit